tools/pmd_test.bzl

```diff
@@ -7,38 +7,71 @@ def _impl(ctx):
         fail("Expecting a single java library")
 
     src_jar = lib.source_jars[0]
-    full_compile_jars = ":".join([f.path for f in lib.full_compile_jars.to_list()])
+    jar_deps = lib.full_compile_jars
+    full_compile_jars = ":".join([f.short_path for f in jar_deps.to_list()])
 
     # Presumably this needs to be updated to actually run the resulting shell script.
-    cmd = "@//tools:pmd"
 
-    args = [
-        cmd,
+    pmd_exe_file = ctx.attr._pmd[DefaultInfo].files_to_run.executable
+    ruleset = ctx.file.ruleset
+    pmd_cmd_args = [
+        pmd_exe_file.short_path,
         "-f text",
-        "-R pmd-ruleset.xml",
+        "-R {}".format(ruleset.short_path),
         "-d {}".format(src_jar.short_path),
-	"-auxclasspath {}".format(full_compile_jars),
+        "-auxclasspath {}".format(full_compile_jars),
+    ]
+    pmd_cmd = " ".join(pmd_cmd_args)
+    script = [
+        "#!/usr/bin/env bash",
+        "set -o errexit",
+        "set -o xtrace",
+        pmd_cmd,
     ]
-    argline = " ".join(args)
 
-    script = '\n'.join([
-        argline,
-    ])
-    print(script)
+    script_content = "\n".join(script)
+    print(script_content)
 
     # Write the file, it is executed by 'bazel test'.
+    test_file = ctx.actions.declare_file(ctx.attr.name)
```

Comment:
Do we need this? The [Skylark example here](https://github.com/bazelbuild/examples/blob/b901a7c7dcb46377b8e72f73e6c9d1a81c009424/rules/test_rule/line_length.bzl#L23) just used `ctx.output.executable` directly.

Does using the name allow for better introspection of the generated files?

Reply:
From the doc you linked elsewhere:

> As legacy behavior, executable rules have a special `ctx.outputs.executable` predeclared output. This file serves as the default executable if you do not specify one using `DefaultInfo`; it must not be used otherwise. This output mechanism is deprecated because it does not support customizing the executable file’s name at analysis time.

I guess we should not follow that example :)